### PR TITLE
Fix Gaps in Views compared to VScode. 

### PR DESCRIPTION
**Pull request by ariel-bentu**
This PR fixes some of the gaps described in https://github.com/theia-ide/theia/issues/4912

Specifically:
1. When a VSCode extension contributes a view, and registers it using vscode.window.registerTreeDataProvider, the dataProvider implements a getChildren returning vscode.TreeItem[].

each such TreeItem may provide iconPath attribute, which may be a string of an object {light:string, dark:string}. Before this change, the iconPath was not making any effect.

Remark: 
From VSCode examples, the way the iconPath is constructed starts from the absolute path of the ts file. for example: 
path.join(__filename, '..', '..', 'resources', 'light', 'bugs.svg')
in my laptop it results : /private/var/folders/hm/zvsqr3ns2d98t7__1v64hk580000gn/T/vscode-unpacked/bugs-0.0.1.vsix/extension/resources/light/bugs.svg
In order for it to work, it needs to be: hostedPlugin/<plugin-id>/resources/light/bugs.svg.

I did not find a clean way to convert the absolute path in the iconPath to the relative path to which I add hostedPlugin/<plugin-id>/
What I currently rely on, is the fact that /extension/ is in the path. better suggestions are welcome.

2. a contributed menu, with the path 'view/item/context' may have a group, and the group named 'inline' is special, as it means the menu item would be rendered as an icon after the caption of the tree item and be clickable to execute the command behind the menu. So now it works.

remark: in vscode, it only renders such menu item as icon and removes it from the  context menu. I did not (yet?) implement that, so it will also render as part of the context menu.

### CODE REVIEW

**Comment by akosyakov** [comment]
@ariel-bentu Should https://github.com/ariel-bentu/theia-vscode-gaps be used for vscode extension?

File: packages/core/src/browser/tree/tree-decorator.ts [hunk 1]
```diff
@@ -282,6 +282,19 @@ export namespace TreeDecoration {
         readonly color?: Color;
     }
 
+    export interface TailDecorationCommand extends BaseTailDecoration {
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-decorator.ts]: @kittaakos Could you review decorator part please?

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-decorator.ts]: 👍 Sure.

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx [hunk 2]
```diff
@@ -19,6 +19,9 @@ import { MAIN_RPC_CONTEXT, TreeViewsMain, TreeViewsExt } from '../../../api/plug
 import { RPCProtocol } from '@theia/plugin-ext/src/api/rpc-protocol';
 import { ViewRegistry } from './view-registry';
 import { Message } from '@phosphor/messaging';
+import { MenuModelRegistry, CompositeMenuNode, ActionMenuNode } from '@theia/core/lib/common/menu';
+import * as path from 'path';
```
Comment (kittaakos) [inline packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: You cannot use Node.js on the frontend.

File: packages/core/src/browser/style/tree.css [hunk 3]
```diff
@@ -108,6 +108,18 @@
...
 
+.theia-TreeNode .theia-TreeNodeTailInlineCommand  {
+    visibility: hidden;
+}
+
+.theia-TreeNode:hover .theia-TreeNodeTailInlineCommand  {
+    visibility: visible;
+}
+
+.theia-TreeNode.theia-mod-selected .theia-TreeNodeTailInlineCommand {
```
Comment (kittaakos) [inline packages/core/src/browser/style/tree.css]: Tree decorator styles should go the `tree-decorators.css`, and the new constants should go under [`Styles`](https://github.com/theia-ide/theia/blob/756bc5d182c10b1af46045f3211c302f20192a29/packages/core/src/browser/tree/tree-decorator.ts#L171).

**Comment by ariel-bentu** [comment]
> @ariel-bentu Should https://github.com/ariel-bentu/theia-vscode-gaps be used for testing?

Yes this is what I also used.
Thanks

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 4]
```diff
@@ -571,10 +576,25 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
                 const { tooltip } = decoration;
                 const { data, fontData } = decoration as TreeDecoration.TailDecoration;
                 const color = (decoration as TreeDecoration.TailDecorationIcon).color;
-                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
-                const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
+                const iconClass = (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
```
Comment (kittaakos) [inline packages/core/src/browser/tree/tree-widget.tsx]: It seems we do not handle the `TreeDecoration.TailDecorationIcon` anymore.

**Comment by ariel-bentu** [comment]
Hi,
Try dark theme, I think I did not put thw bug.svg in the light folder of
the resources in my example extension.


בתאריך יום ג׳, 23 באפר׳ 2019, 09:20, מאת Akos Kitta ‏<
notifications@github.com>:

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 4 above)
Comment (ariel-bentu) [reply packages/core/src/browser/tree/tree-widget.tsx]: fixed with latest commit

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx (hunk 2 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: removed

File: packages/core/src/browser/style/tree.css (hunk 3 above)
Comment (ariel-bentu) [reply packages/core/src/browser/style/tree.css]: done

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 4 above)
Comment (ariel-bentu) [reply packages/core/src/browser/tree/tree-widget.tsx]: > Instead of introducing yet another interface, can we make TailDecorationIconClass and TailDecorationIcon optionally "clickable"?

I started this way, but then I learned that for command with icon, a dynamic css class is created. So the semantic of the iconClass is different. in the TailDecorationIcon and TailDecorationIconClass the class is FA based and in the command case, it is dynamic pointing to the url provided by the extension. Therefore I switched to own interface.

Comment (ariel-bentu) [reply packages/core/src/browser/tree/tree-widget.tsx]: > I have tried your changes and your extension in Gitpod, although it seemed to work, I could see the following error in the logs:

can you instruct me how to do the same? so I can try to debug it in the same environment - in gitpod?

File: packages/plugin-ext/src/common/plugin-protocol.ts [hunk 5]
```diff
@@ -462,6 +462,7 @@ export interface ViewContainer {
 export interface View {
     id: string;
     name: string;
+    urlBase: string;
```
Comment (akosyakov) [inline packages/plugin-ext/src/common/plugin-protocol.ts]: Could you elaborate why do we need it? In other cases we translate icon path to url and transfer it, so code encapsulated in one place. Is there a good reason to do it differently here?

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx [hunk 6]
```diff
@@ -271,6 +359,30 @@ export class TreeViewWidget extends TreeWidget {
         }
     }
 
+    getIconPath(icon: string | { light: string; dark: string }): string {
+        let iconPath: string;
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: Can we handle it in a similar way how we handle command icons: https://github.com/theia-ide/theia/blob/e710223dd3ee8acb70569a3cf8b5266574f47386/packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts#L160-L173

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 4 above)
Comment (kittaakos) [reply packages/core/src/browser/tree/tree-widget.tsx]: - I have opened your PR in Gitpod: https://gitpod.io/#https://github.com/theia-ide/theia/pull/4969
 - Create the `plugins` folder and clone your example there: `mkdir plugins && cd plugins && git clone https://github.com/ariel-bentu/theia-vscode-gaps`
 - Add `"skipLibCheck": true` to the `tsconfig.json` and run `npm i && npm run compile` for your extension.
 - Start Theia: `cd ../../ && yarn --cwd ./examples/browser start`.
 - Click on `Open Browser`.

Screen Shot 2019-04-24 at 08 17 46

 - You can see the contributions; no errors so far.
 - Refersh your browser <kbd>Ctrl/Cmd</kbd>+<kbd>R</kbd>. If asked `Reload site?`, then `Reload`.
 - Now you can see the error, but the decorator icons are visible: `GET https://3000-bd0da029-e6dc-46dd-9902-0785ae55bea4.ws-eu0.gitpod.io/workspace/theia/plugins/theia-vscode-gaps/resources/light/bugs.svg 404`

Comment (kittaakos) [reply packages/core/src/browser/tree/tree-widget.tsx]: > I learned that for command with icon, a dynamic css class is created

I do not have much experience with VS Code extensions; are you saying the `TailDecorationCommand` interface can be used with VS Code extensions only?

File: packages/plugin-ext/src/common/plugin-protocol.ts (hunk 5 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/common/plugin-protocol.ts]: The iconUrl is recieved from a code the vscode-extension developer writes. In MSFT examples, the set it relative to the file in which the code is in. See here https://github.com/ariel-bentu/theia-vscode-gaps/blob/master/src/BugsTree.ts - line 45.
or here https://github.com/Microsoft/vscode-extension-samples/blob/master/tree-view-sample/src/nodeDependencies.ts - line 103.

this comes much after we parse the package.json of the extension... so we need to keep context of which view from which extension so when we get it, we can translate it.

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx (hunk 6 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: I am not a css expert, but again, this url is provided by an interface implemented by the extension and comes late after css is loaded (can be when you expand a tree folder for example.

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 4 above)
Comment (ariel-bentu) [reply packages/core/src/browser/tree/tree-widget.tsx]: Ok, problem was that I used vsix of my example and you used it extracted (git clone), apparently my problematic logic (as pointed out in the pr) to convert absolute url to relative to hostedPlugin/plugin_id/... broke. I now added to the view meta data also the absolute extension path so the logic is now robust and makes no assumption.

A  for using it not in vscode extension, I suppose you mean theia extension? same way plugin-ext can use it, others can too.

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx (hunk 6 above)
Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: It is the same for commands. This code dynamically inserts new css rules and updates it based on theme changes.

File: packages/plugin-ext/src/common/plugin-protocol.ts (hunk 5 above)
Comment (akosyakov) [reply packages/plugin-ext/src/common/plugin-protocol.ts]: Does it mean that we pass absolute fs paths to the client and then try to apply them to this url base? I'm a bit reluctant to reveal extension location to clients, but maybe there is no other way.

BTW could you help me to understand why you check for `/extensions/` in the absolute path?

Comment (ariel-bentu) [reply packages/plugin-ext/src/common/plugin-protocol.ts]: > Does it mean that we pass absolute fs paths to the client
No, the extension developer already has the absolute url using the variable __filename. so the paths to icons are absolute. we need to convert them to rest call to the plugin host, so we need the plugin url (hence the baseUrl) and yesterday I added the absolute base of the extension, so the calculation is robust and does not rely on /extension/ in the path (which was wrong in case it was not vsix but extracted extension for example). the only one using this meta data is the tree-views-main.ts

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx (hunk 6 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: I will try to convert it this way, because as you said, the advantage is that it is sensitive to theme change.
but a question: 
Since tree-nodes are dynamic (and potentially each has different icon) this may produce many rules - is that an issue?

Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: Ok I did. see latest commit.

File: packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts [hunk 7]
```diff
@@ -164,8 +164,8 @@ export class PluginContributionHandler {
                 const darkIconUrl = typeof iconUrl === 'object' ? iconUrl.dark : iconUrl;
                 const lightIconUrl = typeof iconUrl === 'object' ? iconUrl.light : iconUrl;
                 this.style.insertRule('.' + iconClass, theme => `
-                    width: 16px;
-                    height: 16px;
+                    width: 22px;
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts]: How is it related to this PR? Did you check how icons look like in editor toolbars and so on?

File: packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts [hunk 8]
```diff
@@ -126,7 +126,8 @@ export class MenusContributionPointHandler {
             return uri ? uri['codeUri'] : (typeof selection !== 'object' && typeof selection !== 'function') ? selection : undefined;
         };
         this.commands.registerCommand(command, {
-            execute: () => this.commands.executeCommand(menu.command, selectedResource()),
+            // tslint:disable-next-line:no-any
+            execute: (args: any[]) => this.commands.executeCommand(menu.command, selectedResource()),
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts]: Please get rid of changes in unrelated files.

File: packages/core/src/browser/tree/tree-decorator.ts [hunk 9]
```diff
@@ -282,6 +283,19 @@ export namespace TreeDecoration {
...
+        /**
+         * icon class
+         */
+        readonly iconClass: string;
+
+        /**
+         * Optional event handler for click on the tail decorator
+         */
+        // tslint:disable-next-line:no-any
+        readonly onClick: any;
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-decorator.ts]: Is there a proper type?

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx [hunk 10]
```diff
@@ -281,11 +375,50 @@ export class TreeViewWidget extends TreeWidget {
...
-        return undefined;
+    private createClassName(iconPath: string, packagePath: string): string {
+        if (iconPath.startsWith(packagePath)) {
+            iconPath = iconPath.substr(packagePath.length);
+        }
+        return iconPath.replace(/\//g, '_').replace(/\./g, '_');
+    }
+
+    private getIconUrl(iconPath: string, packagePath: string, urlBase: string): string {
+        if (iconPath.startsWith(packagePath)) {
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: there is `URI.relative`

**Comment by akosyakov** [comment]
There are several concerns with this PR.
- I don't understand why we need tree decorators here and changes in core. Decorators should be used to allow external contributors to decorate some tree, here the widget is self-contained in the plugin extension. We can just override render function for a node and render icons without causing additional render cycles by updating decorators.
- So far position of an extension was not exposed to the frontend. This PR sends absolute paths revealing this information. I'm not sure how it is important though. It seems to be hard to avoid. cc @svenefftinge

File: packages/plugin-ext/src/main/browser/view/tree-views-main.tsx (hunk 10 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/view/tree-views-main.tsx]: It seems to fit http uri, with scheme check and authority check. Here it is a filesystem path.

File: packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts (hunk 8 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts]: Done

File: packages/core/src/browser/tree/tree-decorator.ts (hunk 9 above)
Comment (ariel-bentu) [reply packages/core/src/browser/tree/tree-decorator.ts]: Yes - fix.

React.MouseEvent

**Comment by ariel-bentu** [comment]
> * I don't understand why we need tree decorators here and changes in core. 
In this PR there are two things:
* icons left to the tree-node - this is done in plugin-ext only, where the tree is entirely owned by the ext.
* context-menu with group=inline are made available with icon on the tree-node. Here, the changes are in the core, because this is where contributors may want to decorate existing trees.
I believe my code so far is not complete, as it only deals with tree created by the extension.

I would try to amend it to include support for contributors decorating existing tree. (first I check if it works in vs-code...
Makes-sense?

> * So far position of an extension was not exposed to the frontend. 
Not exactly correct: extension developer can figure out where the actual file is running from (server-side-node) by using __filename variable see [here](https://nodejs.org/api/globals.html#globals_filename). and I also don't see other way. 

Thanks

File: packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts (hunk 7 above)
Comment (ariel-bentu) [reply packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts]: reverted, moved to this solution specific css

**Comment by ariel-bentu** [comment]
Is it possible to consider including this PR in the coming release - we (SAP) would be happy to get this, as we need it in our use-case.

Thanks,
Ariel
